# AIM ---------------------------------------------------------------------
# the aim of the script is to read in the raw table of counts and compare it with the one automatically generated by nextflow.

# libraries ---------------------------------------------------------------
library("tidyverse")
library("DESeq2")
library("AnnotationHub")
library("ensembldb")
library("tximport")
library("GenomicFeatures")
library("txdbmaker")

# annotation annotationHub ------------------------------------------------
# Create a transcript dataframe annotation
ah <- AnnotationHub()

# Explore the AnnotationHub object
ah

# Query AnnotationHub
human_ens <- query(ah, c("Homo sapiens", "EnsDb"))
human_ens

# explore the full list of the database
mcols(human_ens) %>% data.frame()

# Extract the lates annotation
# AH75011
# AH119325
human_ens <- human_ens[["AH119325"]]

# pull the transcripts and genes id
txdb <- transcripts(human_ens, return.type = "data.frame") %>%
  dplyr::select(tx_id, gene_id)

dim(txdb)

# filter the transcripts
txdb2 <- txdb[grep("ENST", txdb$tx_id),]

dim(txdb2)
head(txdb2)

# Create a gene-level dataframe
genedb <- genes(human_ens, return.type = "data.frame")  %>%
  dplyr::select(gene_id, gene_name)

dim(genedb)
head(genedb)

# Merge the two dataframes together
annotations <- inner_join(txdb2, genedb)

dim(annotations)
head(annotations)

# save the transcript annotation
saveRDS(annotations,"../../out/object/annotation_annotationHub.rds")

# annotation from GTF -----------------------------------------------------
# library(GenomicFeatures)

# Replace "path/to/your/annotation.gtf" with the actual path to your GTF file
gtf_file <- "../../data/genes_ensembl.gtf"

# Create a TranscriptDb object from the GTF file
txdb <- makeTxDbFromGFF(gtf_file, format = "gtf")
keytypes(txdb)

# Extract the transcript-to-gene mapping
keytypes <- keys(txdb, keytype = "TXNAME")
tx2gene <- select(txdb, keys = keytypes, columns = "GENEID", keytype = "TXNAME")

# The 'tx2gene' data frame will have two columns: TXNAME (transcript ID) and GENEID (gene ID)
head(tx2gene)
dim(tx2gene)

# check if there is any gene versioned. In case it is not needed, remove the version.
# also make sure to keep only the ENST features.
# tx2gene2 <- tx2gene %>%
#   mutate(GENEID_no_version = str_replace_all(GENEID,"\\..*$","")) %>%
#   dplyr::filter(str_detect(TXNAME,pattern = "ENST"))

tx2gene %>%
  dplyr::filter(str_detect(GENEID,pattern = "\\."))

# pull the official gene names form the GTF file
LUT_genes <- rtracklayer::import(gtf_file) %>%
  as.data.frame() %>%
  dplyr::select(GENEID = gene_id, SYMBOL = gene_name) %>%
  distinct()

head(LUT_genes)
dim(LUT_genes)

# inner join the annotations to have all the annotations in one place
annotations2 <- inner_join(tx2gene, LUT_genes)
dim(annotations2)
head(annotations2)

saveRDS(annotations2,"../../out/object/annotation_GTF.rds")

# read in the expression data ---------------------------------------------
# sample import from the salmon aboundance data

# save all the samples
samples <- list.files(path = "../../data/salmon_quant", full.names = T, pattern="S")

# Obtain a vector of all filenames including the path
files <- file.path(samples, "quant.sf")

# Since all quant files have the same name it is useful to have names for each element
names(files) <- str_replace(samples, "../../data/salmon_quant/", "")

# import the aboundance
# use the annotationHub source
txi <- tximport(files, type="salmon", tx2gene = annotations, countsFromAbundance="lengthScaledTPM")

# use the GTF source
txi2 <- tximport(files, type="salmon", tx2gene = annotations2, countsFromAbundance="lengthScaledTPM")

# see the structure of the object
str(txi)
str(txi2)

# save the object
saveRDS(txi,"../../out/object/txi_annotationHub.rds")
saveRDS(txi2,"../../out/object/txi_GTF.rds")

# check the attibutes
attributes(txi)
attributes(txi2)

# Look at the counts
txi$counts %>% head()
txi$counts %>% dim()

txi2$counts %>% head()
txi2$counts %>% dim()

# Write the counts to an object and round them to be imported into DESeq2
data <- txi$counts %>% 
  round()

data2 <- txi2$counts %>% 
  round()

# save the raw table of counts
saveRDS(data,"../../out/object/raw_count_salmon_annotationHub.rds")
saveRDS(data2,"../../out/object/raw_count_salmon_GTF.rds")

# compare the table of counts with the automatic object generated by nextflow
load("../../data/deseq2.dds.RData")
data_nextflow <- counts(dds)

# compare the three tables
dim(data_nextflow)

test_nextflow <- data_nextflow %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  pivot_longer(names_to = "samples",values_to = "count",-gene)

test <- data %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  pivot_longer(names_to = "samples",values_to = "count",-gene)

test2 <- data2 %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  pivot_longer(names_to = "samples",values_to = "count",-gene)

df_test1 <- full_join(test_nextflow,test,by = c("gene","samples"),suffix = c(".nextflow",".annotaionHub"))
df_test2 <- full_join(test_nextflow,test2,by = c("gene","samples"),suffix = c(".nextflow",".GTF"))

# the annotationHub full join has some missing values
df_test1 %>%
  dplyr::filter(!complete.cases(.))

# the annotationHub processing has some deviation from the nextflow output
df_test1 %>%
  mutate(delta = count.nextflow - count.annotaionHub) %>%
  dplyr::filter(delta != 0)

df_test1 %>%
  mutate(delta = count.nextflow - count.annotaionHub) %>%
  # dplyr::filter(delta != 0) %>%
  ggplot(aes(x = count.nextflow,y = count.annotaionHub)) + geom_point() +
  scale_y_continuous(trans = "log1p") +
  scale_x_continuous(trans = "log1p")

# the GTF full join has no missing values
df_test2 %>%
  dplyr::filter(!complete.cases(.))

# the GTF processing are exactly the same values
df_test2 %>%
  mutate(delta = count.nextflow - count.GTF) %>%
  dplyr::filter(delta != 0)

df_test2 %>%
  mutate(delta = count.nextflow - count.GTF) %>%
  # dplyr::filter(delta != 0) %>%
  ggplot(aes(x = count.nextflow,y = count.GTF)) + geom_point() +
  scale_y_continuous(trans = "log1p") +
  scale_x_continuous(trans = "log1p")

# the GTF approach produce exactlry the dame results presented in the nextflow output
